File: s3/s3-client-annotation-processor/src/test/java/ru/tinkoff/kora/s3/client/annotation/processor/S3AwsClientTests.java
package ru.tinkoff.kora.s3.client.annotation.processor;

import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.annotation.processor.common.AbstractAnnotationProcessorTest;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class S3AwsClientTests extends AbstractAnnotationProcessorTest {

    @Override
    protected String commonImports() {
        return super.commonImports() + """
            import java.util.List;
            import java.util.Collection;
            import ru.tinkoff.kora.s3.client.annotation.*;
            import ru.tinkoff.kora.s3.client.annotation.S3.*;
            import ru.tinkoff.kora.s3.client.model.*;
            import ru.tinkoff.kora.s3.client.*;
            import ru.tinkoff.kora.s3.client.model.S3Object;
            import software.amazon.awssdk.services.s3.model.*;
            """;
    }

    @Test
    public void clientGetAws() {
        this.compile(List.of(new S3ClientAnnotationProcessor()), """
            @S3.Client("my")
            public interface Client {
                        
                @S3.Get
                GetObjectResponse get(String key);
            }
            """);
        this.compileResult.assertSuccess();
        var clazz = this.compileResult.loadClass("$Client_Impl");
        assertThat(clazz).isNotNull();
    }

    @Test
    public void clientListAws() {
        this.compile(List.of(new S3ClientAnnotationProcessor()), """
            @S3.Client("my")
            public interface Client {
                        
                @S3.List
                ListObjectsV2Response list();
            }
            """);
        this.compileResult.assertSuccess();
        var clazz = this.compileResult.loadClass("$Client_Impl");
        assertThat(clazz).isNotNull();
    }

    @Test
    public void clientListAwsWithPrefix() {
        this.compile(List.of(new S3ClientAnnotationProcessor()), """
            @S3.Client("my")
            public interface Client {
                        
                @S3.List
                ListObjectsV2Response list(String prefix);
            }
            """);
        this.compileResult.assertSuccess();
        var clazz = this.compileResult.loadClass("$Client_Impl");
        assertThat(clazz).isNotNull();
    }

    @Test
    public void clientListAwsLimit() {
        this.compile(List.of(new S3ClientAnnotationProcessor()), """
            @S3.Client("my")
            public interface Client {
                        
                @S3.List(limit = 100)
                ListObjectsV2Response list(String prefix);
            }
            """);
        this.compileResult.assertSuccess();
        var clazz = this.compileResult.loadClass("$Client_Impl");
        assertThat(clazz).isNotNull();
    }

    @Test
    public void clientDeleteAws() {
        this.compile(List.of(new S3ClientAnnotationProcessor()), """
            @S3.Client("my")
            public interface Client {
                        
                @S3.Delete
                DeleteObjectResponse delete(String key);
            }
            """);
        this.compileResult.assertSuccess();
        var clazz = this.compileResult.loadClass("$Client_Impl");
        assertThat(clazz).isNotNull();
    }

    @Test
    public void clientDeletesAws() {
        this.compile(List.of(new S3ClientAnnotationProcessor()), """
            @S3.Client("my")
            public interface Client {
                        
                @S3.Delete
                DeleteObjectsResponse delete(List<String> key);
            }
            """);
        this.compileResult.assertSuccess();
        var clazz = this.compileResult.loadClass("$Client_Impl");
        assertThat(clazz).isNotNull();
    }

    @Test
    public void clientPutBody() {
        this.compile(List.of(new S3ClientAnnotationProcessor()), """
            @S3.Client("my")
            public interface Client {
                        
                @S3.Put
                PutObjectResponse put(String key, S3Body body);
            }
            """);
        this.compileResult.assertSuccess();
        var clazz = this.compileResult.loadClass("$Client_Impl");
        assertThat(clazz).isNotNull();
    }
}

File: experimental/s3-client-symbol-processor/src/test/kotlin/ru/tinkoff/kora/s3/client/symbol/processor/S3AwsClientTests.kt
package ru.tinkoff.kora.s3.client.symbol.processor

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.ksp.common.AbstractSymbolProcessorTest
import ru.tinkoff.kora.ksp.common.KotlinCompilation

class S3AwsClientTests : AbstractSymbolProcessorTest() {

    override fun commonImports(): String {
        return super.commonImports() + """
            import java.util.List;
            import java.util.Collection;
            import ru.tinkoff.kora.s3.client.annotation.*;
            import ru.tinkoff.kora.s3.client.annotation.S3.*;
            import ru.tinkoff.kora.s3.client.model.*;
            import ru.tinkoff.kora.s3.client.*;
            import ru.tinkoff.kora.s3.client.model.S3Object;
            import software.amazon.awssdk.services.s3.model.*;
            """.trimIndent()
    }

    fun compile0(vararg src: String) = KotlinCompilation()
        .withPartialClasspath()
        .withClasspathJar("s3")
        .withClasspathJar("aws-core")
        .withClasspathJar("sdk-core")
        .withClasspathJar("utils")
        .withClasspathJar("http-client-spi")
        .withClasspathJar("reactor-core")
        .withClasspathJar("reactive-streams")
        .compile(listOf(S3ClientSymbolProcessorProvider()), *src)

    @Test
    fun clientGetAws() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.Get
                fun get(key: String): GetObjectResponse
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientGetOptionalAws() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.Get
                fun get(key: String): GetObjectResponse?
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientGetMetaAws() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.Get
                fun get(key: String): HeadObjectResponse
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientGetMetaOptionalAws() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.Get
                fun get(key: String): HeadObjectResponse?
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientListAws() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.List
                fun list(): ListObjectsV2Response
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientListAwsWithPrefix() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.List
                fun list(prefix: String): ListObjectsV2Response
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientListAwsLimit() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.List(limit = 100)
                fun list(prefix: String): ListObjectsV2Response
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientListKeyAndDelimiter() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.List(value = "some/path/to/{key1}/object", delimiter = "/")
                fun list(key1: String): ListObjectsV2Response
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientDeleteAws() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.Delete
                fun delete(key: String): DeleteObjectResponse
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientDeletesAws() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.Delete
                fun delete(key: List<String>): DeleteObjectsResponse
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }

    @Test
    fun clientPutBody() {
        this.compile0(
            """
            @S3.Client("my")
            interface Client {
                        
                @S3.Put
                fun put(key: String, value: S3Body): PutObjectResponse
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()
        val clazz = loadClass("\$Client_Impl")
        assertThat(clazz).isNotNull()
    }
}
